server/aesdsocket.c

```diff
@@ -239,120 +436,131 @@ int main ( int argc, char **argv )
-            if (length == -1)
-            {
-                syslog(LOG_ERR, "Receive failed");
-                free(buf);
-                close(new_fd);
-                syslog(LOG_DEBUG, "Closed connection from %s", client_ip);
-                continue;
-            }
+            syslog(LOG_ERR, "Thread creation failed");
+            free(server_params);
+            server_params = NULL;
+            continue;
+        }
 
-            total_received += length;
-            end_packet = strchr(buf, '\n');
-            count++;
+        /* Add the node to the SLIST*/
+        SLIST_INSERT_HEAD(&head, server_params, link);
 
-        } while ((end_packet == NULL) && (length > 0));
+        /* Attempt to join threads by checking for the complete_thread flag*/
+        server_thread_params_t *iterator = NULL;
+        server_thread_params_t *tmp = NULL;
 
-    /* Check if we found a newline */
-        if (end_packet != NULL)
-        {
-            /* Append to file only until \n */
-            valid_size = end_packet - buf + 1; 
-            /* Null terminate the string */
-            buf[valid_size] = '\0';
-            fwrite(buf, sizeof(char), valid_size, tmp_file);
-            fflush(tmp_file);
-        }
-        else
-        {
-            syslog(LOG_ERR, "Received without newline\n");
-        }
-            
-        free(buf);
-
-        /* Send back to client */
-        fseek(tmp_file, 0, SEEK_SET);
+        SLIST_FOREACH_SAFE(iterator, &head, link, tmp) 
+	    {
+            if (iterator->thread_complete == true) 
+	        {
+                if(pthread_join(iterator->thread_id, NULL) != 0)
+		        {
+                    syslog(LOG_ERR, "Thread join failed fir %ld", iterator->thread_id);
+                }
+                syslog(LOG_INFO, "Thread joined %ld", iterator->thread_id);
 
-        size_t send_buf_size = valid_size;
-        char *send_buf = (char *)malloc(send_buf_size);
-        if (send_buf == NULL)
-        {
-            syslog(LOG_ERR, "Memory allocation failed for sending buffer");
-            close(new_fd);
-            syslog(LOG_DEBUG, "Closed connection from %s", client_ip);
-            continue;
+                /* Remove node from the list and free the memory */
+                SLIST_REMOVE(&head, iterator, server_thread_params, link);
+                free(iterator);
+                iterator = NULL;
+            }
         }
+    }
 
-        while (fgets(send_buf, send_buf_size, tmp_file) != NULL) 
+    /* Cleanup after caught signal */
+    /* Mutex */
+    pthread_mutex_destroy(&tmp_file_write_mutex);
+    /* Timestamp thread */
+    pthread_cancel(time_params->thread_id);
+    pthread_join(time_params->thread_id, NULL);
+    free(time_params);
+    /* Server thread */
+    server_thread_params_t *iterator = NULL;
+    server_thread_params_t *tmp = NULL;
+    SLIST_FOREACH_SAFE(iterator, &head, link, tmp) 
+    {
+        if (iterator->thread_complete == true) 
```

Comment:
@traptibalgi I think you'll have a race condition here between thread completion and cleanup, and you won't cleanup after a thread which is still running at this point.  Would be better to unconditionally pthread_join here.